zkevm-circuits/src/evm_circuit/execution/calldataload.rs

```diff
@@ -0,0 +1,381 @@
+use std::convert::TryInto;
+
+use eth_types::{Field, ToLittleEndian};
+use halo2_proofs::{
+    circuit::Region,
+    plonk::{Error, Expression},
+};
+
+use crate::{
+    evm_circuit::{
+        param::{N_BYTES_MEMORY_ADDRESS, N_BYTES_WORD},
+        step::ExecutionState,
+        table::{CallContextFieldTag, TxContextFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition},
+            memory_gadget::BufferReaderGadget,
+            Cell, MemoryAddress, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+
+use super::ExecutionGadget;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallDataLoadGadget<F> {
+    /// Gadget to constrain the same context.
+    same_context: SameContextGadget<F>,
+    /// Transaction id from the tx context.
+    tx_id: Cell<F>,
+    /// The bytes offset in calldata, from which we load a 32-bytes word.
+    calldata_start: MemoryAddress<F>,
+    /// Start reading into buffer from this source address.
+    src_addr: Cell<F>,
+    /// End of the source address.
+    src_addr_end: Cell<F>,
+    /// Gadget to read from tx calldata, which we validate against the word
+    /// pushed to stack.
+    buffer_reader: BufferReaderGadget<F, N_BYTES_WORD, N_BYTES_MEMORY_ADDRESS>,
+}
+
+impl<F: Field> ExecutionGadget<F> for CallDataLoadGadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CALLDATALOAD;
+
+    const NAME: &'static str = "CALLDATALOAD";
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let calldata_start = cb.query_rlc();
+
+        // Pop the offset value from stack.
+        cb.stack_pop(calldata_start.expr());
+
+        // Add a lookup constrain for TxId in the RW table.
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        let src_addr = cb.query_cell();
```

Comment:
hmm, there's some problem here. `src_addr` should be either `calldata_start` (root call) or `calldata_start + call_data_offset` (internal call, where `call_data_offset` can be queried from call context table, see [here](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs#L81-L86) in calldatacopy case). And we don't need to allocate a cell for `src_addr` since we can use expression to derive it.